nucypher/blockchain/eth/interfaces.py

```diff
@@ -478,7 +480,12 @@ def build_contract_transaction(self,
                                      use_pending_nonce=use_pending_nonce)
         self.__log_transaction(transaction_dict=payload, contract_function=contract_function)
         try:
-            transaction_dict = contract_function.buildTransaction(payload)  # Gas estimation occurs here
+            if 'gas' not in payload:
+                # As web3 buildTransaction() will estimate gas with block identifier "pending" by default,
+                # explicitly estimate gas here with block identifier 'latest' if not otherwise specified
+                # as a pending transaction can cause gas estimation to fail, notably in case of worklock refunds.
+                payload['gas'] = contract_function.estimateGas(payload, 'latest')
```

Comment:
:+1:

Reply:
This works with recent geth versions but it's worth checking whether estimateGas() with a block identifier parameter works when using another node (old geth, parity, infura) or if it throws (as suggested by @cygnusv I believe).

Reply:
Indeed it does appear that Infura supports the `eth_estimateGas` endpoint (@cygnusv) - https://infura.io/docs/ethereum#operation/eth_estimateGas